src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorLanguageMiddleware.ts

```diff
@@ -0,0 +1,100 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { getRazorDocumentUri, isRazorCSharpFile } from './RazorConventions';
+import { RazorLanguageServiceClient } from './RazorLanguageServiceClient';
+import { RazorLogger } from './RazorLogger';
+import { LanguageKind } from './RPC/LanguageKind';
+
+// This interface should exactly match the `LanguageMiddleware` interface defined in Omnisharp.
+interface LanguageMiddleware {
+
+    language: string;
+
+    remapWorkspaceEdit?(workspaceEdit: vscode.WorkspaceEdit, token: vscode.CancellationToken): vscode.ProviderResult<vscode.WorkspaceEdit>;
+
+    remapLocations?(locations: vscode.Location[], token: vscode.CancellationToken): vscode.ProviderResult<vscode.Location[]>;
+}
+
+export class RazorLanguageMiddleware implements LanguageMiddleware {
+    public readonly language = 'Razor';
+
+    constructor(
+        private readonly serviceClient: RazorLanguageServiceClient,
+        private readonly logger: RazorLogger) {}
+
+    public async remapWorkspaceEdit(workspaceEdit: vscode.WorkspaceEdit, token: vscode.CancellationToken) {
```

Comment:
I don't want this to block this PR because we need to get it in soon but we should write unit tests for these remappings (`Microsoft.AspNetCore.Razor.VSCode.Test`)